### PR TITLE
CCIP-1703 Offramp Execution State Changes - Testing finality

### PR DESCRIPTION
**Pull request by mateusz-sekara**
## Context 
https://github.com/smartcontractkit/ccip/pull/566

### CODE REVIEW

File: core/services/ocr2/plugins/ccip/internal/ccipdata/v1_0_0/offramp_reader_unit_test.go [hunk 1]
```diff
@@ -154,3 +157,81 @@ func generateTokensAndOutputs(nbTokens uint) ([]common.Address, []common.Address
...
+				lpLogs[i] = logpoller.Log{
+					BlockNumber: log.blockNumber,
+				}
+				gethLogs[i] = lpLogs[i].ToGethLog()
+			}
+
+			lp.On("IndexedLogsTopicRange", mock.Anything, mock.Anything, mock.Anything, mock.Anything, mock.Anything, mock.Anything, mock.Anything).
+				Return(lpLogs, nil)
+
+			mockOffRamp := mock_contracts.NewEVM2EVMOffRampInterface(t)
```
Comment (mateusz-sekara) [inline core/services/ocr2/plugins/ccip/internal/ccipdata/v1_0_0/offramp_reader_unit_test.go]: I couldn't make `ExecutionStateChange` events serialize properly to the format which `contract.ParseExecutionStateChanged` can deserialize easily

Comment (makramkd) [reply core/services/ocr2/plugins/ccip/internal/ccipdata/v1_0_0/offramp_reader_unit_test.go]: Is that important for this test? Don't think so right?

**Review by makramkd — DISMISSED** [review]

**Comment by makramkd** [comment]
Oops, I hastily approved. Tests are still failing. Having a look.

File: core/services/ocr2/plugins/ccip/internal/ccipdata/v1_0_0/offramp.go [hunk 2]
```diff
@@ -451,7 +451,7 @@ func (o *OffRamp) GetExecutionStateChangesBetweenSeqNums(ctx context.Context, se
 
 			return &cciptypes.ExecutionStateChanged{
 				SequenceNumber: sc.SequenceNumber,
-				Finalized:      sc.Raw.BlockNumber <= uint64(latestBlock.FinalizedBlockNumber),
+				Finalized:      sc.Raw.BlockNumber >= uint64(latestBlock.FinalizedBlockNumber),
```
Comment (makramkd) [inline core/services/ocr2/plugins/ccip/internal/ccipdata/v1_0_0/offramp.go]: Oh, intentional test failure? 😄

Comment (mateusz-sekara) [reply core/services/ocr2/plugins/ccip/internal/ccipdata/v1_0_0/offramp.go]: Yes, this was intentional to verify if test is valid

File: core/services/ocr2/plugins/ccip/internal/ccipdata/v1_0_0/offramp_reader_unit_test.go (hunk 1 above)
Comment (mateusz-sekara) [reply core/services/ocr2/plugins/ccip/internal/ccipdata/v1_0_0/offramp_reader_unit_test.go]: No, it's not, but usually the less mocking the more maintainable the test is

Comment (makramkd) [reply core/services/ocr2/plugins/ccip/internal/ccipdata/v1_0_0/offramp_reader_unit_test.go]: If you're still wondering how to do it you can do something like this:

```golang
tAbi, err := evm_2_evm_offramp.EVM2EVMOffRampMetaData.GetAbi()
require.NoError(t, err)
eseEvent, ok := tAbi.Events["ExecutionStateChanged"]
require.True(t, ok)

seqNr := uint64(12)
logData, err := eseEvent.Inputs.NonIndexed().Pack(uint8(1), []byte("some return data"))
require.NoError(t, err)
seqNrBytes := make([]byte, 8)
binary.BigEndian.PutUint64(seqNrBytes, seqNr)
topic0 := evm_2_evm_offramp.EVM2EVMOffRampExecutionStateChanged{}.Topic()
topic1 := common.BytesToHash(seqNrBytes)
topic2 := crypto.Keccak256Hash([]byte("some message id"))

// use topic0, topic1 and topic2 as the topics
// use topic0 as the event sig
// use logData as the log.Data
lg := logpoller.Log{
	Topics: [][]byte{
		topic0[:],
		topic1[:],
		topic2[:],
	},
	Data:        logData,
	LogIndex:    1,
	BlockHash:   common.HexToHash("0x1234"),
	BlockNumber: 1234,
	EventSig:    topic0,
	Address:     testutils.NewAddress(),
	TxHash:      common.HexToHash("0x123456"),
}
wrapper, err := evm_2_evm_offramp.NewEVM2EVMOffRamp(lg.Address, nil)
require.NoError(t, err)
eseParsed, err := wrapper.ParseExecutionStateChanged(lg.ToGethLog())
require.NoError(t, err)
require.Equal(t, uint8(1), eseParsed.State)
require.Equal(t, []byte("some return data"), eseParsed.ReturnData)
require.Equal(t, topic2[:], eseParsed.MessageId[:])
```

Comment (mateusz-sekara) [reply core/services/ocr2/plugins/ccip/internal/ccipdata/v1_0_0/offramp_reader_unit_test.go]: Whoa, nice, let me check if that works

Comment (makramkd) [reply core/services/ocr2/plugins/ccip/internal/ccipdata/v1_0_0/offramp_reader_unit_test.go]: Ah actually I was slightly incorrect, you have to include topic0 in the logpoller topics. Update the snippet above.

File: core/services/ocr2/plugins/ccip/internal/ccipdata/v1_0_0/test_helpers.go [hunk 3]
```diff
@@ -42,3 +47,32 @@ func ApplyPriceRegistryUpdate(t *testing.T, user *bind.TransactOpts, addr common
...
 	require.NoError(t, err)
 }
+
+func CreateExecutionStateChangeEventLog(t *testing.T, seqNr uint64, blockNumber int64, messageID common.Hash) logpoller.Log {
+	tAbi, err := evm_2_evm_offramp.EVM2EVMOffRampMetaData.GetAbi()
+	require.NoError(t, err)
+	eseEvent, ok := tAbi.Events["ExecutionStateChanged"]
+	require.True(t, ok)
+
+	logData, err := eseEvent.Inputs.NonIndexed().Pack(uint8(1), []byte("some return data"))
```
Comment (makramkd) [inline core/services/ocr2/plugins/ccip/internal/ccipdata/v1_0_0/test_helpers.go:57]: Do you care about the non-indexed arguments here? I believe the first one is the execution state and the second is `returnData`.

**Review by makramkd — APPROVED** [review]

File: core/services/ocr2/plugins/ccip/internal/ccipdata/v1_0_0/test_helpers.go (hunk 3 above)
Comment (mateusz-sekara) [reply core/services/ocr2/plugins/ccip/internal/ccipdata/v1_0_0/test_helpers.go:57]: They are not relevant for the test, but let's keep them, might be useful in the future

**Pull request merged**